### ISSUE TITLE
Cannot instantiate WASM built with -s USE_PTHREADS=1 on mobile device

### ISSUE THREAD

**Issue by erestor** [emscripten-core/emscripten#9598]
(Seen in Chrome on Android on Samsung Galaxy S7.)

I expect a module built with thread support to instantiate in a browser, even if SharedArrayBuffer is not defined, and run correctly unless the multi-threading functionality is actually invoked. The current instantiation fallbacks from SharedArrayBuffer to ArrayBuffer and tests in pthread library for presence of SharedArrayBuffer lead me to think this is the expected funtionality.

This is a bit more complicated so let me elaborate:
The first problem when loading the pthread-enabled wasm module on mobile is a plain JS error `SharedArrayBuffer undefined` in the assert in [runtime_init_memory.js, line 38](https://github.com/emscripten-core/emscripten/blob/incoming/src/runtime_init_memory.js#38). (Note the assertion itself doesn't execute, because it uses the undefined symbol.)
When I fix that and bypass the assertion, I hit `Atomics undefined` in `initMainThreadBlock` in [library_pthread.js, line 60](https://github.com/emscripten-core/emscripten/blob/incoming/src/library_pthread.js#34). I fix that by inserting a no-op pattern for no SharedArrayBuffer support seen in e.g. `allocateUnusedWorkers` to the top of the function.
Then I finally hit the real problem which I can't fix, which is the instantiation failure.

Warnings in the console:
* `wasm streaming compile failed: CompileError: WebAssembly.instantiateStreaming(): invalid memory limits flags`
* `falling back to ArrayBuffer instantiation`
* `failed to asynchronously prepare wasm: CompileError: WebAssembly.instantiate(): invalid memory limits flags`
* `CompileError: WebAssembly.instantiate(): invalid memory limits flags`
And finally this leads to the error
* `Uncaught (in promise) abort(CompileError: WebAssembly.instantiate(): invalid memory limits flags`

I tried changing the memory creation flag in [runtime_init_memory.js](https://github.com/emscripten-core/emscripten/blob/incoming/src/runtime_init_memory.js#34) to `'shared': typeof SharedArrayBuffer !== 'undefined'` which didn't help.

I tried this with `-s ALLOW_MEMORY_GROWTH=1 -s WASM_MEM_MAX=512Mb` and I also tried this with `-s TOTAL_MEMORY=256Mb`, it makes no difference, so clearly the 'invalid flags' are something else.

I also traced the 'invalid memory limits flags' message to [wasm module decoder](https://github.com/nodejs/node/blob/master/deps/v8/src/wasm/module-decoder.cc), but that's about as far as I can see. I don't know where these flags come from.

Reproduce:
Open [skolaris.net/dev](https://skolaris.net/dev) in Chrome on mobile. Remote debugging necessary of course to see the warnings/errors in the console. Note that my two workarounds for the javascript issues mentioned above have been applied here to reveal the instantiation failure itself.
(Open the same page on a computer - works fine.)

Any clues, @juj @VirtualTim ?

**Comment by kripken** [comment]
I think mobile browsers don't yet have threads support enabled. It's only on desktop, and on Chrome, that it's enabled by default now.

There might be a pref you can flip, I'm not sure.

**Comment by VirtualTim** [comment]
Yeah @kripken is right. To expand on this, the Emscripten threads are implemented on top of SharedArrayBuffer. This is the mechanism through which data is transferred between threads.

The error "Wasm decoding failed: invalid memory limits flags" basically means that the browser doesn't have threading support. Like @kripken said, that's only by default on Chromium based browsers (Chrome, Opera, Edge beta), but you can also enable it in Firefox by setting "javascript.options.shared_memory".

So long story short, you need a browser that support SharedArrayBuffer to use a `USE_PTHREADS` build. [On mobile it currently that looks like it's default enabled on UC Browser](https://caniuse.com/#feat=sharedarraybuffer), but you might be able to set on Chrome: chrome://flags/#enable-webassembly-threads. From the description that might work on Android?

**Comment by erestor** [comment]
Guys I'm sorry but can you actually read the issue again? I'm not asking for SharedArrayBuffer support in mobile browsers, but a graceful fallback to ArrayBuffer and successful _instantiation_ of the module so that I can use WASM functions that do not use multi threading.

**Comment by VirtualTim** [comment]
Ah, sorry, my mistake.
I guess that should be possible, but I think you're probably better off creating two separate builds anyway. USE_PTHREADS generally works best with PTHREAD_POOL_SIZE (better performance, less chance of deadlocks), and in non-pthreads builds you can do things like use different MALLOC's.

Looking at places like https://github.com/emscripten-core/emscripten/blob/3a142bf374063597fc8b3fa51f7718eccd41760d/src/library_pthread.js#L274
makes me think it should at least load.

If you compile something like:
```
include <cstdio>
int main() {
  printf("loaded!\n");
}
```
with -s USE_PTHREADS=1
does that work?

**Comment by AjayP13** [comment]
@erestor I did the same thing as you actually, inserting no-op patterns for no SharedArrayBuffer support to support running a browser that did not have SharedArrayBuffer. And I failed at the compile error as well. 

I believe the error I got was something more specific from Desktop Firefox that said something like: "CompileError: WebAssembly.instantiate(): shared memory not enabled".

Seemed to me at that point that the `.wasm` binary compiled with `-s USE_PTHREADS=1` is made to work only with memory that is shared so even if you don't intend to use pthreads, it won't instantiate unless the underlying buffer of the `WebAssembly.Memory` is a SharedArrayBuffer. I think you might have to build the `.wasm` binary twice (once with pthreads and once without) and then swap based on whether SharedArrayBuffers are enabled or not unfortunately.

**Comment by AjayP13** [comment]
It would be nice if a graceful fallback could work though. It makes it so that you can't really compile with `-s USE_PTHREADS=1` since no environment supports this today other than Chrome. It would be nice if the non-threaded functionality could still work in other browsers by running on an ArrayBuffer instead of SharedArrayBuffer.

But I'm not sure if this is a limitation of Emscripten or of WebAssembly.

**Comment by kripken** [comment]
Yes, there isn't a way to fall back to non-pthreads mode. A wasm binary compiled with pthreads support will have a Memory that declared itself as shared, and atomic instructions, all of which a wasm VM without threads support will error on.

In theory we could modify the binary on the client, removing those things before the VM sees it, but parsing and unparsing is quite slow. Two separate builds is the best option we have for now. However, there is discussion in the standards bodies about "fat binaries" for wasm that would contain both pthreads and non-pthreads code, with automatic fallback, so eventually this may get better.

**Comment by erestor** [comment]
Thank you all.

**Comment by delleceste** [comment]
Hello. How is the status of Chrome + Webassembly + Threads on ANDROID?
When I try to load this
https://www.qt.io/web-assembly-example-mandelbrot

I get 
Qt for WebAssembly:  RangeError: WebAssembly.Memory():

I get the error even for a very small project with a single C++-11 thread incrementing an integer displayed on a text label.

Am I missing something?

In chrome://flags I ENABLED WebAssembly threads support.

Thanks. 

Giacomo

**Comment by kripken** [comment]
@delleceste I believe it's only on desktop for now. But there is ongoing work towards that.

**Comment by VirtualTim** [comment]
I have a pthreads build that runs on Android.

The error: `WebAssembly: RangeError: WebAssembly.Memory():`
Could be because too much memory is being allocated in one go. `new WebAssembly.Memory` requires continuous memory, which it might not be able to find on mobile.

What `TOTAL_MEMORY` are you compiling with?
Can you try running something like this in the console: `memory_test = new WebAssembly.Memory({initial:10000});  //10000 =10k*64KB ~=640MB` where 10000 is however many 64KB pages you're trying to allocate.
